### PR TITLE
[WIP] Add a very basic Makie extension

### PR DESCRIPTION
**Pull request by asinghvi17**
## Checklist

- [ ] Appropriate tests were added
- [X] Any code changes were done in a way that does not break public API
- [ ] All documentation related to code changes were updated
- [x] The new code follows the
  [contributor guidelines](https://github.com/SciML/.github/blob/master/CONTRIBUTING.md), in particular the [SciML Style Guide](https://github.com/SciML/SciMLStyle) and
  [COLPRAC](https://github.com/SciML/COLPRAC).
- [ ] Any new documentation only uses public API
  
## Additional context

I thought I'd update the code in #427 for the latest Makie, and while I was at it made it an extension as well. 

This doesn't work fully yet, in that:
1. It is only hooked up to AbstractTimeseriesSolutions.
2. It is currently only able to plot line plots.
3. It currently does not accept theme keywords passed to `plot`, like `linewidth`.  Setting that theme externally still works, though.

### CODE REVIEW

File: ext/SciMLBaseMakieExt.jl [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+                !(hasfield(typeof(sol), :interp) &&
+                    typeof(sol.interp) <: SciMLBase.SensitivityInterpolation),
+    plotdensity = min(Int(1e5),
+                        sol.tslocation == 0 ?
+                        (sol.prob isa SciMLBase.AbstractDiscreteProblem ?
+                        max(1000, 100 * length(sol)) :
+                        max(1000, 10 * length(sol))) :
+                        1000 * sol.tslocation),
+    plotat = nothing,
+    tspan = nothing, # we don't support this, since this is a plot recipe, not an axis spec!
```
Comment (ChrisRackauckas) [inline ext/SciMLBaseMakieExt.jl:34]: wouldn't you just return a different size array?

**Comment by asinghvi17** [comment]
Ok, made some decent progress here!  The recipes are now to the point that they can plot any DiffEq solution to any point-based plot type, depending on what the user provides.  Color cycling also comes naturally and is affected by theme.

On the Makie level, it's currently not supported to pass themes from the plot _invocation_ to subsidiary recipes unless they are specified and consumed, so you can't adjust things like `linewidth` in the `plot` call (though you can do that in the global theme, which does trickle down).  Once that's solved on the Makie end it should be a relatively simple update here.

Here are some examples

<details>
  <summary>Code here</summary>
  
```julia
using DifferentialEquations, Makie
using CairoMakie

function lorenz(du,u,p,t)
    du[1] = p[1]*(u[2]-u[1])
    du[2] = u[1]*(p[2]-u[3]) - u[2]
    du[3] = u[1]*u[2] - p[3]*u[3]
end

u0 = [1., 5., 10.]
tspan = (0., 100.)
p = (10.0,28.0,8/3)
prob = ODEProblem(lorenz, u0, tspan,p)
sol = solve(prob)

f, a, p = plot(sol)
axislegend(a)
f

f = Figure()
topgl = GridLayout(f[1, 1])
botgl = GridLayout(f[2, 1])
xyzt, _p = plot(topgl[1, 1:2], sol, plotdensity=10000,linewidth=1.5)
xy, _p = plot(topgl[1, 3], sol, plotdensity=10000, idxs=(1,2))
xz, _p = plot(botgl[1, 1], sol, plotdensity=10000, idxs=(1,3))
yz, _p = plot(botgl[1, 2], sol, plotdensity=10000, idxs=(2,3))
xyz, _p = plot(botgl[1, 3], sol, plotdensity=10000, idxs=(1,2,3))

axislegend.((xyzt, xy, xz, yz, xyz))
f

_f(x,y,z) = (sqrt(x^2+y^2+z^2),x)
plot(sol,idxs=(_f,1,2,3))
sol

f = Figure()
lines(f[1, 1], sol)
scatter(f[2, 1], sol; denseplot = false)
scatterlines(f[3, 1], sol; denseplot = false, plotdensity = 1)
f

function fx(du, u, p, t)
    du[1] = p[1] * u[1] - p[2] * u[1] * u[2]
    du[2] = -3 * u[2] + u[1] * u[2]
end

function gx(du, u, p, t)
    du[1] = p[3] * u[1]
    du[2] = p[4] * u[2]
end

p = [1.5, 1.0, 0.1, 0.1]
prob = SDEProblem(fx, gx, [1.0, 1.0], (0.0, 10.0), p)

prob_func = let p = p
    (prob, i, repeat) -> begin
        x = 0.3rand(2)
        remake(prob, p = [p[1], p[2], x[1], x[2]])
    end
end

ensemble_prob = EnsembleProblem(prob, prob_func = prob_func)
sim = solve(ensemble_prob, SRIW1(), trajectories = 10)

plot(sim)
plot(sim; idxs = (0, 1))

summ = EnsembleSummary(sim, 0:0.1:10)

plot(summ)
```
</details>

A basic plot
download-4
Support for a random recipe which is defined as consuming point-based input
download-5
Theming :)
download-6
Ensemble
download-1
Ensemble with errorbars
download

**Comment by ChrisRackauckas** [comment]
> I also added basic support for ensembles, which seem to be the only things which had special treatment in SciMLBase's Plots recipes so far.

Integrators have a recipe as well

**Comment by ChrisRackauckas** [comment]
I think this looks good?

**Comment by asinghvi17** [comment]
It should work for most simple usecases - more advanced ones will probably need to be deconstructed manually, but that's not particularly hard...

**Comment by ChrisRackauckas** [comment]
That's fine, it's a start. We can get it in there and let it keep improving.

**Comment by henry2004y** [comment]
I have a question about the recipe, but I am not sure if this is the right place to ask @asinghvi17 . Say if I want to scale the axis (e.g. time) by a factor of 10 for plotting, how should I do that? Makie provides a function `scale!`; it works for a simple case like

```julia
f = Figure()
ax = Axis(f[1, 1], aspect=1)
l0 = lines!(ax, [0, 1], [0, 1])
scale!(l0, 10, 1)
```

but not for the DiffEq recipe:

```julia
using OrdinaryDiffEq
using GLMakie

function lorenz(du,u,p,t)
    du[1] = p[1]*(u[2]-u[1])
    du[2] = u[1]*(p[2]-u[3]) - u[2]
    du[3] = u[1]*u[2] - p[3]*u[3]
end

u0 = [1., 5., 10.]
tspan = (0., 100.)
p = (10.0,28.0,8/3)
prob = ODEProblem(lorenz, u0, tspan,p)
sol = solve(prob, Tsit5())

f, a, p = plot(sol)
axislegend(a)
# This is not working!
scale!(p, 10, 1)

f
```

**Comment by asinghvi17** [comment]
Ah! I think this is another issue uncovered - the API I used to do this is still pretty new!

You can of course scale the plots individually, by inspecting e.g. `ax.scene.plots`, but this seems like an issue which has to be fixed in Makie.  The primary thing that has to be solved is that a plot's Transformation should be passed to its children in the event that a plotlist is being generated.

Would you mind filing an issue on the Makie repo, @henry2004y?

**Comment by henry2004y** [comment]
Sure! Hopefully there will be a quick fix. Thanks for the new recipe!

**Comment by henry2004y** [comment]
And another maybe related issue is that the new recipe does not work with plotting attributes (e.g. `linestyle, linewidth`)? Is this an issue for Makie or the recipe @asinghvi17 ?

<details>
  <summary>MWE</summary>
  
```julia
using OrdinaryDiffEq
using GLMakie

function lorenz(du,u,p,t)
    du[1] = p[1]*(u[2]-u[1])
    du[2] = u[1]*(p[2]-u[3]) - u[2]
    du[3] = u[1]*u[2] - p[3]*u[3]
end

u0 = [1., 5., 10.]
tspan = (0., 100.)
p = (10.0,28.0,8/3)
prob = ODEProblem(lorenz, u0, tspan,p)
sol = solve(prob, Tsit5())
f, a, p = lines(sol, linestyle="dot")
```
</details>

**Comment by asinghvi17** [comment]
Same issue unfortunately :(

**Comment by asinghvi17** [comment]
Same issue unfortunately :(

**Comment by henry2004y** [comment]
> Same issue unfortunately :(

Is there a temporary workaround for attributes before it's properly fixed in Makie?

**Comment by henry2004y** [comment]
Honestly I am now pretty frustrated about the new SpecAPI introduced in Makie v0.20. The old approaches for creating recipes did not limit it capability in so many ways.

Here is yet another issue with `PlotSpec`: it does not work (yet) with `lift`:

<details>
  <summary>MWE</summary>

```julia
using OrdinaryDiffEq
using GLMakie

function lorenz(du,u,p,t)
    du[1] = p[1]*(u[2]-u[1])
    du[2] = u[1]*(p[2]-u[3]) - u[2]
    du[3] = u[1]*u[2] - p[3]*u[3]
end

u0 = [1., 5., 10.]
tspan = (0., 100.)
p = (10.0,28.0,8/3)
prob = ODEProblem(lorenz, u0, tspan,p)
sol = solve(prob, Tsit5())

fig = Figure()

ax = Axis(fig[1, 1])

t1, t2 = 0.0, 100.0
step = (t2 - t1) / 1e3
time_format = t2 < 2e-3 ? "{:.3e} s" : "{:.3f} s"
# the minimum of the range cannot be t1.
sg = SliderGrid(fig[2, 1],
    (label="Time",
     range=range(t1+step, stop=t2, step=step),
     format=time_format,
     startvalue=t2))
# convert tspan to an Observable
tspan = lift(sg.sliders[1].value) do t
    (t1, t)
end

lines!(ax, sols; idxs=1, tspan)
# Drag the slider and errors pop out!
```
</details>

Do you think it's possible to switch back to the more stable APIs for now and switch to the newer experimental Makie APIs once it becomes mature? @asinghvi17

**Pull request merged**